server/src/main/java/org/opensearch/index/translog/RemoteFsTranslog.java

```diff
@@ -0,0 +1,245 @@
+        ExecutorService executorService
+    ) throws IOException {
+        super(config, translogUUID, deletionPolicy, globalCheckpointSupplier, primaryTermSupplier, persistedSequenceNumberConsumer);
+        this.blobStoreRepository = blobStoreRepository;
+        fileTransferTracker = new FileTransferTracker(shardId);
+        this.translogTransferManager = new TranslogTransferManager(
+            new BlobStoreTransferService(blobStoreRepository.blobStore(), executorService),
+            blobStoreRepository.basePath().add(shardId.getIndex().getUUID()).add(String.valueOf(shardId.id())),
+            fileTransferTracker,
+            fileTransferTracker::exclusionFilter
+        );
+
+        try {
+            final Checkpoint checkpoint = readCheckpoint(location);
+            this.readers.addAll(recoverFromFiles(checkpoint));
+            if (readers.isEmpty()) {
+                throw new IllegalStateException("at least one reader must be recovered");
+            }
+            boolean success = false;
+            current = null;
+            try {
+                current = createWriter(
+                    checkpoint.generation + 1,
+                    getMinFileGeneration(),
+                    checkpoint.globalCheckpoint,
+                    persistedSequenceNumberConsumer
+                );
+                success = true;
+            } finally {
+                // we have to close all the recovered ones otherwise we leak file handles here
+                // for instance if we have a lot of tlog and we can't create the writer we keep
+                // on holding
+                // on to all the uncommitted tlog files if we don't close
+                if (success == false) {
+                    IOUtils.closeWhileHandlingException(readers);
+                }
+            }
+        } catch (Exception e) {
+            // close the opened translog files if we fail to create a new translog...
+            IOUtils.closeWhileHandlingException(current);
+            IOUtils.closeWhileHandlingException(readers);
+            throw e;
+        }
+    }
+
+    @Override
+    public boolean ensureSynced(Location location) throws IOException {
+        try (ReleasableLock ignored = writeLock.acquire()) {
+            assert location.generation <= current.getGeneration();
+            if (location.generation == current.getGeneration()) {
+                ensureOpen();
+                return prepareAndUpload(primaryTermSupplier.getAsLong(), location.generation);
+            }
+        } catch (final Exception ex) {
+            closeOnTragicEvent(ex);
+            throw ex;
+        }
+        return false;
+    }
+
+    @Override
+    public void rollGeneration() throws IOException {
+        syncBeforeRollGeneration();
+        if (current.totalOperations() == 0 && primaryTermSupplier.getAsLong() == current.getPrimaryTerm()) {
+            return;
+        }
+        prepareAndUpload(primaryTermSupplier.getAsLong(), null);
```

Comment:
While `ensureSynced` uploads within a writeLock is this true even for `rollGeneration`. To ensure exactly one upload can happen at any point in time we need to ensure that the write lock is held by the current thread. Can we add those checks

Reply:
`prepareAndUpload` already does a `writeLock` acquire, which will guarantee that exactly one thread will be uploading the txlog files. Since `Translog` has a `ReentrantReadWriteLock` , it will ensure that only thread will be able to take the write log, and other threads will lie dormant, till the write lock is released .